File: go_example/TtsExample.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/golang/glog"
)

const ttsHttpUrl = "https://open.mobvoi.com/api/tts/v1"

// appKey和appSecret从https://openapi.mobvoi.com/user/mine-app-list创建应用获取
var appKey = ""
var appSecret = ""

var ttsPostData = map[string]interface{}{
	"signature":    "appkey+secret+timestamp的MD5值",
	"timestamp":    "",
	"appkey":       appKey,
	"speaker":      "cissy_meet",
	"ignore_limit": true,
	"gen_srt":      true,
	"audio_type":   "mp3",
	"text":         "海南长臂猿的叫声，高亢洪亮，响彻山谷。海南热带雨林国家公园是这种濒危灵长类动物的全球唯一栖息地。经过近年来的科学保护和生态恢复，海南长臂猿已由最少时的寥寥几只，恢复到5群35只，创造了世界珍稀动物保护的奇迹。国家公>园堪称最美国土，具有典型独特的自然生态系统、世界瞩目的野生动植物种。在海南热带雨林国家公园，这里生长着846种特有植物、145种国家重点保护野生动物，生物多样性指数与巴西亚马孙雨林相当。2018年4月，习近平总书记在庆祝海>南建省办经济特区30周年大会上强调，要积极开展国家公园体制试点，建设热带雨林等国家公园。2019年1月，总书记又主持召开中央全面深化改革委员会第六次会议，审议通过《海南热带雨林国家公园体制试点方案》。被称为海南“生态绿心”的这片最美国土迈出保护和建设的历史性一步。我国的国家公园在自然保护地体系中保护等级最高、生态价值最大、管控措施最严。",
	"speed":        "1.1",
}

func genMD5Sum(s string) string {
	glog.Info("genMD5Sum: ", s)
	md5_hash := md5.New()
	md5_hash.Write([]byte(s))
	ret := md5_hash.Sum(nil)
	glog.V(1).Info("genMD5Sum bytes: ", ret)
	rets := hex.EncodeToString(ret)
	glog.Info("genMD5Sum: ", rets)
	return rets
}

func getTimeNow() int64 {
	now := time.Now()
	return now.Unix()
}

func ttsSample() {
	ttsPostData["timestamp"] = fmt.Sprintf("%v", getTimeNow())
	ttsPostData["signature"] = genMD5Sum(fmt.Sprintf("%v+%v+%v", appKey, appSecret, ttsPostData["timestamp"]))
	// convert ttsPostData to String
	ttsPostByte, err := json.Marshal(ttsPostData)
	if err != nil {
		glog.Error(err)
		return
	}

	ttsPostStr := string(ttsPostByte)
	glog.Info("post data: " + ttsPostStr)

	payload := strings.NewReader(ttsPostStr)
	method := "POST"
	client := &http.Client{}
	req, err := http.NewRequest(method, ttsHttpUrl, payload)

	if err != nil {
		glog.Error(err)
		return
	}
	req.Header.Add("Content-Type", "application/json")

	res, err := client.Do(req)
	if err != nil {
		glog.Error(err)
		return
	}
	glog.Info(res)
	if res.StatusCode != 200 {
		glog.Error("response code: ", res.StatusCode)
		return
	}
	if res.Header.Get("Content-Type") != "audio/mpeg" {
		glog.Warning("Content-Type: ", res.Header.Get("Content-Type"))
		b, _ := io.ReadAll(res.Body)
		glog.Warning("Response body: ", string(b))
		return
	}
	defer res.Body.Close()

	out, err := os.Create("demo.mp3")
	if err != nil {
		panic(err)
	}
	defer out.Close()

	_, err = io.Copy(out, res.Body)
	if err != nil {
		panic(err)
	}
}

func main() {
	flag.Set("logtostderr", "true")
	flag.Set("v", "2")
	flag.Set("stderrthreshold", "INFO")
	ttsSample()
}

File: python3_example/TtsExample.py
#!/usr/bin/env python3
# coding=utf-8
import time
import hashlib
import os
import json
import requests
from loguru import logger

# APP_KEY和APP_SECRET从https://openapi.mobvoi.com/user/mine-app-list创建应用获取
APP_KEY = ""
APP_SECRET = ""
TTS_HTTP_URL = "https://open.mobvoi.com/api/tts/v1"

tts_post_data ={
    'text': '出门问问成立于2012年，是一家以语音交互和软硬结合为核心的人工智能公司，为全球40多个国家和地区的消费者、企业提供人工智能产品和服务。',
    'speaker': 'xiaoyi_meet',
    'audio_type': 'mp3',
    'speed': 1.0,
    #'symbol_sil': 'semi_250,exclamation_300,question_250,comma_200,stop_300,pause_150,colon_200', # 停顿调节需要对appkey授权后才可以使用，授权前传参无效。
    #'ignore_limit': True, # 忽略1000字符长度限制，需要对appkey授权后才可以使用
    'gen_srt': False, # 是否生成srt字幕文件，默认不开启。如果开启生成字幕，需要额外计费。生成好的srt文件地址将通过response header中的srt_address字段返回。
    'appkey': APP_KEY,
    'timestamp': "",
    'signature': ""
}

def gen_md5_signature(message: str) -> str:
    """Generate md5sum of a string message."""
    logger.debug(":" + message)
    md5_hasher = hashlib.md5()
    md5_hasher.update(message.encode("utf-8"))
    signature = md5_hasher.hexdigest()
    logger.debug(message + " md5 value: " + signature)
    return signature

def sample() -> None:
    """Sample to synthesis online tts."""
    timestamp = str(int(time.time()))
    tts_post_data['timestamp'] = timestamp
    message = '+'.join([APP_KEY, APP_SECRET, timestamp])
    tts_post_data['signature'] = gen_md5_signature(message)
    try:
        headers = {'Content-Type': 'application/json'}
        logger.info(f"Post data: {json.dumps(tts_post_data)}")
        ret = requests.post(url=TTS_HTTP_URL, headers=headers,
                            data=json.dumps(tts_post_data), timeout=None)
        content = ret.content
        if ret.status_code != 200:
            logger.error(f"response status code : {ret.status_code}")
            return
        if ret.headers.get('Content-Type') != "audio/mpeg":
            logger.error(f"response invalid Content-Type! {ret.headers}")
            logger.info(f"response body {ret.content.decode()}")
            return
        with open(os.path.join(os.path.dirname(os.path.abspath("__file__")), "sample.mp3"), "wb") as f:
            f.write(content)
    except Exception as e:
        logger.error(f"Exception {e}")

def sample_with_srt() -> None:
    """Sample to synthesis online tts with srt."""
    timestamp = str(int(time.time()))
    tts_post_data['timestamp'] = timestamp
    message = '+'.join([APP_KEY, APP_SECRET, timestamp])
    tts_post_data['signature'] = gen_md5_signature(message)
    tts_post_data['audio_type'] = "wav"
    tts_post_data['gen_srt'] = True
    try:
        headers = {'Content-Type': 'application/json'}
        logger.info(f"Post data: {json.dumps(tts_post_data)}")
        ret = requests.post(url=TTS_HTTP_URL, headers=headers,
                            data=json.dumps(tts_post_data), timeout=None)
        content = ret.content
        if ret.status_code != 200:
            logger.error(f"response status code : {ret.status_code}")
            return
        if ret.headers.get('Content-Type') != "audio/mpeg":
            logger.error(f"response invalid Content-Type! {ret.headers}")
            logger.info(f"response body {ret.content.decode()}")
            return
        with open(os.path.join(os.path.dirname(os.path.abspath("__file__")), "sample_with_srt.mp3"), "wb") as f:
            f.write(content)
        srt_url = ret.headers.get('srt_address', None)
        if srt_url is None:
            logger.warning('not found srt url from response header')
            return
        logger.info('srt url:', srt_url)
        content = requests.get(url=srt_url, timeout=None).content
        with open(os.path.join(os.path.dirname(os.path.abspath("__file__")), "sample_with_srt.srt"), "wb") as f:
            f.write(content)
    except Exception as e:
        logger.error(f"Exception {e}")

def sample_with_ssml() -> None:
    """Sample to synthesis online tts with ssml."""
    timestamp = str(int(time.time()))
    tts_post_data['timestamp'] = timestamp
    message = '+'.join([APP_KEY, APP_SECRET, timestamp])
    tts_post_data['signature'] = gen_md5_signature(message)
    ssml_text = '<speak version="1.0" xml:lang="zh-CN" xmlns="http://www.w3.org/2001/10/synthesis">9月10日，庆祝2019年<w phoneme="jiao4 shi1 jie2">教师节</w>暨全国教育系统先进集体和先进个人表彰大会在京举行。<break time="500ms" />习近平总书记在人民大会堂亲切会见受表彰代表，<break time="500ms" />向受到表彰的先进集体和先进个人表示热烈祝贺，<break time="500ms" />向全国广大<p phoneme="jiao4">教</p>师和教育工作者致以节日的问候。</speak>'
    tts_post_data["text"] = ssml_text
    tts_post_data['audio_type'] = "wav"
    tts_post_data['gen_srt'] = True
    try:
        headers = {'Content-Type': 'application/json'}
        logger.info(f"Post data: {json.dumps(tts_post_data)}")
        ret = requests.post(url=TTS_HTTP_URL, headers=headers,
                            data=json.dumps(tts_post_data), timeout=None)
        content = ret.content
        if ret.status_code != 200:
            logger.error(f"response status code : {ret.status_code}")
            return
        if ret.headers.get('Content-Type') != "audio/mpeg":
            logger.error(f"response invalid Content-Type! {ret.headers}")
            logger.info(f"response body {ret.content.decode()}")
            return
        with open(os.path.join(os.path.dirname(os.path.abspath("__file__")), "sample_with_ssml.mp3"), "wb") as f:
            f.write(content)
    except Exception as e:
        logger.error(f"Exception {e}")

def main():
    sample()
    sample_with_srt()
    sample_with_ssml()

if __name__ == '__main__':
    main()
